providers/gitlab/provider/discovery.go

```diff
@@ -0,0 +1,216 @@
+package provider
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+	"github.com/xanzy/go-gitlab"
+	"go.mondoo.com/cnquery/providers-sdk/v1/inventory"
+	"go.mondoo.com/cnquery/providers/gitlab/connection"
+	"golang.org/x/exp/slices"
+)
+
+func (s *Service) discover(root *inventory.Asset, conn *connection.GitLabConnection) (*inventory.Inventory, error) {
+	if conn.Conf.Discover == nil {
+		return nil, nil
+	}
+
+	client := conn.Client()
+	if client == nil {
+		return nil, nil
+	}
+
+	assets := []*inventory.Asset{}
+	projects := []*gitlab.Project{}
+
+	targets := conn.Conf.Discover.Targets
+
+	// The following calls to discover Groups and Projects will always return
+	// gitlab.Group and gitlab.Project objects, no matter if we connect to only
+	// one system or many. This reduces code complexity.
+
+	groupAssets, groups, err := s.discoverGroups(root, conn)
+	if err != nil {
+		return nil, err
+	}
+	if slices.Contains(targets, DiscoveryGroup) {
+		assets = append(assets, groupAssets...)
+	}
+
+	projectAssets, projects, err := s.discoverProjects(root, conn, groups)
```

Comment:
is this not guarded by an if because its needed for the terraform?

Reply:
yes, and for any other mechanisms we may support in the future. The new code for both `discoverGroups` and `discoverProjects` behaves in a way where it will end up with a list of groups and projects either way!

This means: 
1. if you specify one group and one project, and set discover to groups,projects => you get 1 group and 1 project
2. if you specify one group and no project, and set discovery to projects,terraform => you get N projects and N terraforms
3. if you specify one group and no projects, and set discovery to ONLY terraform => you get N terraforms (only!)